Attempt to make batch system more robust and debuggable

diekhans opened the pull request:
**WARNING: this has not been tests I have no access to test batch systems.  People who use them will have to test them before merging into master.**


Log results from HPC batch system commands before parsing.

Check status of calls to batch system commands.

Add popen_communicate that does both subprocess Popen and communicate and
checks for errors.  Also includes stderr in exception that is raised.

johnbradley commented:
Testing a simple CWL workflow using slurm toil never finished. This job is running a simple CWL tar workflow created by adding the `--noStdOutErr` flag to the with instructions here https://github.com/DataBiosphere/toil/issues/2855#issue-522401082.
In slurm (sacct) I can see that the workflow command (tar) had finished successfully but the logs kept print out a lines like the following again and again:
```
... 2020-02-10 09:29:20,787 MainThread DEBUG toil.batchSystems.slurm: squeue output 10245772_[0%20] PD 0:00
```
I think toil kept trying (and failing) to determine that the job was completed.

johnbradley commented on src/toil/batchSystems/slurm.py:
```diff
@@ -127,26 +130,32 @@ def _getJobDetailsFromScontrol(self, slurmJobID):
                 # slurm_load_jobs error: Invalid job id specified
                 # There is no job information, so exit.
                 if len(values)>0 and values[0] == 'slurm_load_jobs':
                     return (None, None)
```
I think the early exit `return (None, -999)` might be related to [an issue where the toil never finishes](https://github.com/DataBiosphere/toil/pull/2959#issuecomment-584207037) on this branch.
The comment at the bottom seems to say that it should `return (None, None)`.

For completed jobs slurm responds with the following for scontrol:
```
$ scontrol show job 15671080
slurm_load_jobs error: Invalid job id specified
```

diekhans replied:
thanks, yes only _getJobDetailsFromSacct() should return -999 magic number..

diekhans commented:
I fixed the incorrect return and added logging for the commands
being executed.  Perhaps give it another try..

John Bradley <notifications@github.com> writes:
> Testing a simple CWL workflow using slurm toil never finished. This job is running a simple CWL tar workflow created by adding the `--noStdOutErr` flag to the with instructions here https://github.com/DataBiosphere/toil/issues/2855#issue-522401082.
> In slurm (sacct) I can see that the workflow command (tar) had finished successfully but the logs kept print out a lines like the following again and again:
> ```
> ... 2020-02-10 09:29:20,787 MainThread DEBUG toil.batchSystems.slurm: squeue output 10245772_[0%20] PD 0:00
> ```
> I think toil kept trying (and failing) to determine that the job was completed.
> 
> 
> 
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/DataBiosphere/toil/pull/2959#issuecomment-584207037
> Testing a simple CWL workflow using slurm toil never finished. This job is
> running a simple CWL tar workflow created by adding the --noStdOutErr flag to
> the with instructions here #2855 (comment).
> In slurm (sacct) I can see that the workflow command (tar) had finished
> successfully but the logs kept print out a lines like the following again and
> again:
> 
> ... 2020-02-10 09:29:20,787 MainThread DEBUG toil.batchSystems.slurm: squeue output 10245772_[0%20] PD 0:00
> 
> I think toil kept trying (and failing) to determine that the job was completed.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.*
>

johnbradley commented:
I retested and all the scenarios I tried functioned as expected.
Scenarios tested with `--logDebug`:
- toil-cwl-runner on Slurm using a shared workdir - Succeeded
- toil-cwl-runner on Slurm using default workdir (/tmp on the worker node) and specifying `--noStdOutErr` flag  - Succeeded
- toil-cwl-runner on Slurm using node local storeage and without specifying `--noStdOutErr` flag  - Failed as expected but with better logging `sacct output b'FAILED|1:0\n'`

I used the same CWL workflow as before detailed here: https://github.com/DataBiosphere/toil/issues/2855#issue-522401082

adamnovak commented on src/toil/lib/misc.py:
```diff
@@ -243,3 +254,53 @@ def atomic_copyobj(src_fh, dest_path, length=16384):
+    (stdout, stderr). Raise an exception if the process exits non-zero. Kwargs
+    are as passed to subprocess.Popen, with input and timeout being passed to
+    communicate.  The PY3.6 Popen encode and error arguments are allowed and
+    implemented.  Generally, Toil should used encoding='latin1'.
```
Shouldn't we be using `utf8` throughout instead of `latin1`? As soon as someone runs Toil against a grid engine scheduler that titles its columns in Chinese or something, we're going to get in trouble with `latin1`.

diekhans replied:
I am not convinced latin1 is right either ;-)

The problem, which I should have documented, is that we don't want to compound the
error reporting problem by having garbage on stderr cause a decoding exception with an invalid UTF8 character.  latin1 can handle all bytes without error.

Probably better would be to use utf-8 and set error to ignore or replace.

I have kind of decide that this is change is both a bit more code than it needs to be and doesn't go far enough.  I am thinking of making a wrapper around the new subprocess.run function that all of toil will use to call process. This way we  would get logging even of all programs executed plus provide the recommended encoding/error as defaults.

subprocess is somewhat limited in that you really want stderr to be strings even if stdin/stdout is binary ..

sigh, why can't ASCII be the universal language

adamnovak replied:
If we want to handle all possible bytes, it would make sense to be passing around bytes objects; we can repr() them for logging. But yeah, ignoring errors in the decoding would solve the same problem too. I just want to avoid a situation where we have to re-encode the library function's return value and decode it as utf-8 because some batch system prints out happy emoji for completed jobs or something.

diekhans replied:
it now reads as utf-8,  however it also sets locale to C, due to Perl locale issues on one system.

diekhans commented:
thanks @johnbradley, 

this is interesting:
> * toil-cwl-runner on Slurm using node local storeage and without specifying `--noStdOutErr` flag  - Failed as expected but with better logging `sacct output b'FAILED|1:0\n'`

not clear why this is happening, as the exec of sacct should have overridden the redirecting of stdout/error that happens  when running the command.

adamnovak commented:
> not clear why this is happening, as the exec of sacct should have overridden the redirecting of stdout/error that happens when running the command.

I don't think that is true; the standard output/error redirection for the job itself is accomplished by telling Slurm to redirect the streams when it actually gets around to running the job. I think that in this case, doing that still makes the job fail, and then the `sacct` output describing the failed job isn't parsed by Toil, but it is logged.

diekhans commented:
Ah, maybe I misinterpreted @johnbradley description as indicating -noStdOutErr was required to trigger the failure, but maybe that isn't the only variable.

johnbradley commented:
> Ah, maybe I misinterpreted @johnbradley description as indicating -noStdOutErr was required to trigger the failure, but maybe that isn't the only variable.

@diekhans sorry for the confusion. When running toil with the slurm backend and using local storage for workdir the toil worker will likely fail to start. I created issue #2855 for this problem. `sbatch` is the slurm command used in toil to run the toil worker. The directory that should contain the stdout and stderr logs for `sbatch` likely will not exist yet on whatever node the worker is run on. `sbatch` quits immediately if it can't write to it's stdout and stderr logs.

If a user specifies the `--noStdOutErr` flag this failure doesn't happen. For a user to figure that out with the current logging wasn't easy. With the new logging changes you have made here it is easer to see where the failure is happening. The `sacct` command displays the accounting/historical database for slurm jobs. The `sacct` command is working fine and reporting that `sbatch` has previously failed with the slurm exit code of `1:0`.

mr-c commented:
I can test on LSF. @johnbradley what is your testing regime?

mr-c commented on src/toil/lib/misc.py:
```diff
@@ -10,11 +13,19 @@
 from contextlib import contextmanager
 
+# can't do from 'toil import subprocess' as __init__.py import this module!
+if os.name == 'posix' and six.PY2:
```
Aren't we py3 only now?

diekhans replied:
this was done before py3 only; I am going to update and simplify

johnbradley commented:
> I can test on LSF. @johnbradley what is your testing regime?

@mr-c See my notes above https://github.com/DataBiosphere/toil/pull/2959#issuecomment-584639194

mr-c commented:
I'm running this branch with the CWL v1.0 conformance tests on LSF, it looks fine to me

adamnovak commented:
I merged master into this branch; that should fix the missing-pytz problem.

diekhans commented:
also fixes #3027

diekhans commented:
I have made several changes to this PR to simplify it and improve logging  GridEngine (UGE) seems to be working well on a system I happen to have access to.

If others could try out that current version in other environments, that would be most helpful.

diekhans commented:
fixes  #3027
